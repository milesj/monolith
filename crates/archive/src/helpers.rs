use moon_error::{map_io_to_fs_error, MoonError};
use moon_utils::path;
use std::fs;
use std::path::{Path, PathBuf};

pub fn ensure_dir(dir: &Path) -> Result<(), MoonError> {
    if !dir.exists() {
        fs::create_dir_all(dir).map_err(|e| map_io_to_fs_error(e, dir.to_path_buf()))?;
    }

    Ok(())
}

pub fn prepend_name(name: &str, prefix: &str) -> String {
    if prefix.is_empty() {
        return name.to_owned();
    }

    // Use native path utils to join the paths, so we can ensure
    // the parts are joined correctly within the archive!
    let parts: PathBuf = [prefix, name].iter().collect();

    path::normalize(parts).to_string_lossy().to_string()
}
