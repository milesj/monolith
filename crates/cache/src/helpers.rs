use moon_logger::warn;
use std::env;
use std::time::SystemTime;

pub const LOG_TARGET: &str = "moon:cache";

static mut LOGGED_WARNING: bool = false;

pub fn get_cache_env_var() -> String {
    if let Ok(var) = env::var("MOON_CACHE") {
        if var == "off" || var == "read" || var == "write" {
            return var;
        }

        // We only want to show this once, not everytime the function is called
        unsafe {
            if !LOGGED_WARNING {
                LOGGED_WARNING = true;

                warn!(
                    target: LOG_TARGET,
                    "Unknown MOON_CACHE environment variable value \"{}\", falling back to write mode",
                    var
                );
            }
        }
    }

    String::from("write")
}

pub fn is_readable() -> bool {
    get_cache_env_var() != "off"
}

pub fn is_writable() -> bool {
    get_cache_env_var() == "write"
}

pub fn to_millis(time: SystemTime) -> u128 {
    match time.duration_since(SystemTime::UNIX_EPOCH) {
        Ok(d) => d.as_millis(),
        Err(_) => 0,
    }
}

#[cfg(test)]
pub async fn run_with_env<T, F, Fut>(env: &str, callback: F) -> T
where
    F: FnOnce() -> Fut,
    Fut: std::future::Future<Output = T>,
{
    if env.is_empty() {
        env::remove_var("MOON_CACHE");
    } else {
        env::set_var("MOON_CACHE", env);
    }

    let result = callback().await;

    env::remove_var("MOON_CACHE");

    result
}
