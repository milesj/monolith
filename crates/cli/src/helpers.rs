use console::{set_colors_enabled, set_colors_enabled_stderr};
use indicatif::{ProgressBar, ProgressStyle};
use moon_terminal::create_theme;
use starbase_styles::color::{no_color, supports_color};
use std::env;
use std::time::Duration;

pub fn map_list<T, F>(list: &[T], fmt: F) -> String
where
    F: Fn(&T) -> String,
{
    list.iter().map(fmt).collect::<Vec<_>>().join(", ")
}

pub fn fully_qualify_version(version: &str) -> String {
    if version.is_empty() {
        return version.to_owned();
    }

    let mut parts = version.split('.');

    [
        parts.next().unwrap_or("0"),
        parts.next().unwrap_or("0"),
        parts.next().unwrap_or("0"),
    ]
    .join(".")
}

pub fn create_progress_bar<S: AsRef<str>, F: AsRef<str>>(start: S) -> impl FnOnce(F, bool) {
    let pb = ProgressBar::new_spinner();
    pb.set_message(start.as_ref().to_owned());
    pb.enable_steady_tick(Duration::from_millis(50));

    move |finish, passed| {
        let theme = create_theme();

        pb.set_style(
            ProgressStyle::default_spinner()
                .template("{prefix} {msg}")
                .unwrap(),
        );

        if passed {
            pb.set_prefix(theme.success_prefix.to_string());
        } else {
            pb.set_prefix(theme.error_prefix.to_string());
        }

        pb.finish_with_message(finish.as_ref().to_owned());
    }
}

fn setup_no_colors() {
    env::set_var("NO_COLOR", "1");
    // https://github.com/mitsuhiko/clicolors-control/issues/19
    env::set_var("CLICOLOR", "0");

    set_colors_enabled(false);
    set_colors_enabled_stderr(false);
}

pub fn setup_colors(force: bool) {
    // If being forced by --color or other env vars
    if force
        || env::var("MOON_COLOR").is_ok()
        || env::var("FORCE_COLOR").is_ok()
        || env::var("CLICOLOR_FORCE").is_ok()
    {
        let mut color_level = env::var("MOON_COLOR")
            .or_else(|_| env::var("FORCE_COLOR"))
            .unwrap_or("3".to_owned());

        // https://nodejs.org/api/cli.html#force_color1-2-3
        if color_level.is_empty() || color_level == "true" {
            color_level = "1".to_owned();
        } else if color_level == "false" {
            color_level = "0".to_owned();
        }

        if color_level == "0" {
            setup_no_colors();
        } else {
            set_colors_enabled(true);
            set_colors_enabled_stderr(true);

            // https://bixense.com/clicolors/
            env::set_var("CLICOLOR_FORCE", &color_level);
            env::set_var("FORCE_COLOR", &color_level);
        }

        return;
    }

    if no_color() {
        setup_no_colors();
    } else {
        env::set_var("CLICOLOR", supports_color().to_string());
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serial_test::serial;

    fn reset_vars() {
        env::remove_var("NO_COLOR");
        env::remove_var("CLICOLOR");
        env::remove_var("CLICOLOR_FORCE");
        env::remove_var("FORCE_COLOR");
        env::remove_var("MOON_COLOR");
    }

    mod setup_color {
        use super::*;

        mod no_color {
            use super::*;

            #[test]
            #[serial]
            fn sets_vars() {
                env::set_var("NO_COLOR", "1");

                setup_colors(false);

                assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                reset_vars();
            }
        }

        mod forced_color {
            use super::*;

            // #[test]
            // #[serial]
            // fn forces_via_arg() {
            //     setup_colors(true);

            //     assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "2");
            //     assert_eq!(env::var("FORCE_COLOR").unwrap(), "2");
            //     assert!(env::var("NO_COLOR").is_err());

            //     reset_vars();
            // }

            // #[test]
            // #[serial]
            // fn forces_over_no_color() {
            //     env::set_var("NO_COLOR", "1");

            //     setup_colors(true);

            //     assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "2");
            //     assert_eq!(env::var("FORCE_COLOR").unwrap(), "2");
            //     assert_eq!(env::var("NO_COLOR").unwrap(), "1");

            //     reset_vars();
            // }

            #[test]
            #[serial]
            fn disables_if_zero() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "0");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                    assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn disables_if_false_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "false");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR").unwrap(), "0");
                    assert_eq!(env::var("NO_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn enables_if_empty_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "1");
                    assert_eq!(env::var("FORCE_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }

            #[test]
            #[serial]
            fn enables_if_true_string() {
                for var in ["MOON_COLOR", "FORCE_COLOR"] {
                    env::set_var(var, "true");

                    setup_colors(false);

                    assert_eq!(env::var("CLICOLOR_FORCE").unwrap(), "1");
                    assert_eq!(env::var("FORCE_COLOR").unwrap(), "1");

                    reset_vars();
                }
            }
        }
    }
}
