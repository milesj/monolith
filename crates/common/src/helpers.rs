use std::sync::OnceLock;

pub fn enable_pkl_configs() {
    std::env::set_var("MOON_EXPERIMENT_PKL_CONFIG", "true");
}

pub fn supports_pkl_configs() -> bool {
    static PKL_CACHE: OnceLock<bool> = OnceLock::new();

    *PKL_CACHE.get_or_init(|| {
        std::env::var("MOON_EXPERIMENT_PKL_CONFIG")
            .is_ok_and(|value| value == "1" || value == "true")
    })
}
