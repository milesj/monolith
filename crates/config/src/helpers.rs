use crate::errors::ConfigError;
use moon_utils::{
    fs::{self, temp},
    path,
};
use std::collections::VecDeque;
use std::path::{Path, PathBuf};
use yaml_rust::{Yaml, YamlLoader};

pub fn download_and_cache_config(url: &str) -> Result<PathBuf, ConfigError> {
    let file = temp::get_file(url, "yml");

    if temp::read(&file)?.is_some() {
        return Ok(file);
    }

    let error_handler = |_| ConfigError::FailedDownload(url.to_owned());
    let data = reqwest::blocking::get(url)
        .map_err(error_handler)?
        .text()
        .map_err(error_handler)?;

    temp::write(&file, data)?;

    Ok(file)
}

pub fn gather_extended_sources<T: AsRef<Path>>(
    base_config: T,
) -> Result<Vec<PathBuf>, ConfigError> {
    let base_config = base_config.as_ref();
    let mut queue = VecDeque::from(vec![path::to_string(base_config)?]);
    let mut sources = vec![];

    while !queue.is_empty() {
        if let Some(config) = queue.pop_front() {
            let mut config_path = PathBuf::from(&config);

            if !config.ends_with(".yml") && !config.ends_with(".yaml") {
                return Err(ConfigError::UnsupportedExtendsDocument(config));
            }

            // For https, download the config and store it in the temp cache
            if config.starts_with("https") {
                config_path = download_and_cache_config(&config)?;
            } else if config.starts_with("http") {
                return Err(ConfigError::UnsupportedHttps(config));
            }

            // Otherwise we have a possible file path
            if !config_path.exists() {
                return Err(ConfigError::MissingFile(config));
            }

            sources.push(config_path.clone());

            // Parse the YAML document and attempt to extract the `extends` field.
            // We can't use serde here as the shape of the document may be invalid
            // or incomplete.
            let yaml = YamlLoader::load_from_str(&fs::sync::read(&config_path)?)
                .map_err(|e| ConfigError::InvalidYaml(config_path.clone(), e.to_string()))?;

            let doc = &yaml[0];

            // Field does not exist!
            if doc["extends"].is_badvalue() {
                continue;
            }

            match &doc["extends"] {
                Yaml::String(extends) => {
                    if extends.starts_with("http") {
                        queue.push_back(extends.to_owned());
                    } else {
                        queue.push_back(path::to_string(
                            config_path.parent().unwrap().join(extends),
                        )?);
                    }
                }
                _ => {
                    return Err(ConfigError::InvalidExtendsField);
                }
            }
        }
    }

    // Reverse the order as we must load the extended leaf first,
    // with the initial base config overriding as last.
    sources.reverse();

    Ok(sources)
}
