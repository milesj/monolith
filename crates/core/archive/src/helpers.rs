use moon_utils::path;
use std::path::PathBuf;

pub fn prepend_name(name: &str, prefix: &str) -> String {
    if prefix.is_empty() {
        return name.to_owned();
    }

    // Use native path utils to join the paths, so we can ensure
    // the parts are joined correctly within the archive!
    let parts: PathBuf = [prefix, name].iter().collect();

    path::normalize(parts).to_string_lossy().to_string()
}
