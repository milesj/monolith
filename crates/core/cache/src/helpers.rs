use moon_logger::warn;
use std::env;

pub const LOG_TARGET: &str = "moon:cache";

static mut LOGGED_WARNING: bool = false;

pub enum CacheLevel {
    Off,
    Read,
    ReadWrite,
    Write,
}

impl From<String> for CacheLevel {
    fn from(value: String) -> Self {
        match value.to_lowercase().as_str() {
            "off" => CacheLevel::Off,
            "read" => CacheLevel::Read,
            "read-write" => CacheLevel::ReadWrite,
            "write" => CacheLevel::Write,
            val => {
                // We only want to show this once, not everytime the function is called
                unsafe {
                    if !LOGGED_WARNING {
                        LOGGED_WARNING = true;

                        warn!(
                            target: LOG_TARGET,
                            "Unknown MOON_CACHE environment variable value \"{}\", falling back to read-write mode",
                            val
                        );
                    }
                }

                CacheLevel::ReadWrite
            }
        }
    }
}

impl CacheLevel {
    pub fn is_readable(&self) -> bool {
        matches!(&self, CacheLevel::Read | CacheLevel::ReadWrite)
    }

    pub fn is_writable(&self) -> bool {
        matches!(&self, CacheLevel::Write | CacheLevel::ReadWrite)
    }
}

pub fn get_cache_level() -> CacheLevel {
    if let Ok(var) = env::var("MOON_CACHE") {
        return CacheLevel::from(var);
    }

    CacheLevel::ReadWrite
}
