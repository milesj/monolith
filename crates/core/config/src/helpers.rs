use crate::errors::ConfigError;
use moon_utils::{
    fs::temp,
    path, time,
    yaml::{self, YamlValue},
};
use std::collections::VecDeque;
use std::path::{Path, PathBuf};

pub fn download_and_cache_config(url: &str) -> Result<PathBuf, ConfigError> {
    let file = temp::get_file(url, "yml");

    if temp::read(&file)?.is_some() {
        return Ok(file);
    }

    let error_handler = |_| ConfigError::FailedDownload(url.to_owned());
    let data = reqwest::blocking::get(url)
        .map_err(error_handler)?
        .text()
        .map_err(error_handler)?;

    temp::write(
        &file,
        format!(
            "# source: {}\n# timestamp: {}\n\n{}",
            url,
            time::now_timestamp(),
            data
        ),
    )?;

    Ok(file)
}

pub fn gather_extended_sources<T: AsRef<Path>>(
    base_config: T,
) -> Result<Vec<PathBuf>, ConfigError> {
    let base_config = base_config.as_ref();
    let mut queue = VecDeque::from(vec![path::to_string(base_config)?]);
    let mut sources = vec![];

    while !queue.is_empty() {
        if let Some(config) = queue.pop_front() {
            let mut config_path = PathBuf::from(&config);

            if !config.ends_with(".yml") && !config.ends_with(".yaml") {
                return Err(ConfigError::UnsupportedExtendsDocument(config));
            }

            // For https, download the config and store it in the temp cache
            if config.starts_with("https") || config.contains("127.0.0.1") {
                config_path = download_and_cache_config(&config)?;
            } else if config.starts_with("http") {
                return Err(ConfigError::UnsupportedHttps(config));
            }

            // Otherwise we have a possible file path
            if !config_path.exists() {
                return Err(ConfigError::MissingFile(config));
            }

            sources.push(config_path.clone());

            // Parse the YAML document and attempt to extract the `extends` field.
            // We can't use serde here as the shape of the document may be invalid
            // or incomplete.
            let doc: YamlValue = yaml::read(&config_path)
                .map_err(|e| ConfigError::InvalidYaml(config_path.clone(), e.to_string()))?;

            if let Some(extends_field) = doc.get("extends") {
                match &extends_field {
                    YamlValue::String(extends) => {
                        if extends.starts_with("http") {
                            queue.push_back(extends.to_owned());
                        } else {
                            queue.push_back(path::to_string(
                                config_path.parent().unwrap().join(extends),
                            )?);
                        }
                    }
                    _ => {
                        return Err(ConfigError::InvalidExtendsField);
                    }
                }
            }
        }
    }

    // Reverse the order as we must load the extended leaf first,
    // with the initial base config overriding as last.
    sources.reverse();

    Ok(sources)
}
