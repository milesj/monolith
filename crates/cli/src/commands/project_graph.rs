use crate::helpers::{generate_project_graph, load_workspace};

pub async fn project_graph(project_id: &Option<String>) -> Result<(), Box<dyn std::error::Error>> {
    let mut workspace = load_workspace().await?;
    let project_graph = generate_project_graph(&mut workspace).await?;

    if let Some(id) = project_id {
        project_graph.get(id)?;
    } else {
        project_graph.get_all()?;
    }

    println!("{}", project_graph.to_dot());

    Ok(())
}
