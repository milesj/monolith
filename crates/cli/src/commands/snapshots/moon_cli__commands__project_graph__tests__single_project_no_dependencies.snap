---
source: crates/cli/src/commands/project_graph.rs
assertion_line: 58
expression: get_assert_output(&assert)

---
digraph {
    0 [ label="(workspace)" style=filled, shape=circle, fillcolor=black, fontcolor=white]
    1 [ label="baz" style=filled, shape=circle, fillcolor=gray, fontcolor=black]
    0 -> 1 [ arrowhead=none]
}


